File: backend/src/main/java/com/uip/backend/forecast/ForecastResult.java
package com.uip.backend.forecast;

import java.time.Instant;
import java.util.Collections;
import java.util.List;

/**
 * Forecast result DTO — trả về từ ForecastPort implementations.
 *
 * @param tenantId     tenant context
 * @param buildingId   building context
 * @param model        model name ("ARIMA", "NAIVE", etc.)
 * @param isFallback   true if naive fallback was used
 * @param mape         MAPE score (null if not computed)
 * @param points       forecast data points
 * @param generatedAt  when this forecast was generated
 */
public record ForecastResult(
        String tenantId,
        String buildingId,
        String model,
        boolean isFallback,
        Double mape,
        List<ForecastPoint> points,
        Instant generatedAt
) {
    public static ForecastResult insufficientData(String tenantId, String buildingId) {
        return new ForecastResult(
                tenantId, buildingId, "NONE", true, null,
                Collections.emptyList(), Instant.now()
        );
    }
}

File: applications/forecast-service/models/forecast_result.py
from dataclasses import dataclass, field
from datetime import datetime
from typing import Optional


@dataclass
class ForecastPoint:
    timestamp: datetime
    actual_value: Optional[float] = None
    predicted_value: float = 0.0
    confidence_upper: float = 0.0
    confidence_lower: float = 0.0
    is_anomaly: bool = False


@dataclass
class ForecastResult:
    tenant_id: str
    building_id: str
    model: str = "ARIMA"
    is_fallback: bool = False
    mape: Optional[float] = None
    points: list[ForecastPoint] = field(default_factory=list)
    generated_at: Optional[datetime] = None
